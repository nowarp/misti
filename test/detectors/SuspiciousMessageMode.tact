contract SendParametersTestContract {
    // Correct usage: should not trigger any warnings
    fun correctUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: SendRemainingValue | SendPayGasSeparately // Ok
        });
    }

    // Duplicate flag usage: should trigger a warning about flags used multiple times
    fun duplicateFlagUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: SendRemainingValue | SendRemainingValue // Bad
        });
    }

    // Invalid operator usage: should trigger a warning about using '+' instead of '|'
    fun invalidOperatorUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: SendRemainingValue + SendIgnoreErrors // Bad
        });
    }

    // Integer literal usage: should trigger a warning about using integer literals
    fun integerLiteralUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: 64 // Bad: Integer literal instead of symbolic constant 
        });
    }

    // Correct combination usage: should not trigger any warnings
    fun correctCombinationUsage() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: SendRemainingBalance | SendDestroyIfZero | SendIgnoreErrors // Ok
        });
    }

    // Multiple issues: should trigger multiple warnings
    fun multipleIssues() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: SendRemainingValue + SendRemainingValue + 64 // Bad: Duplicate flags, '+' operator, integer literal 
        });
    }

    // Complex expression with nested invalid operators and unary operation
    fun complexInvalidOperator() {
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: -(SendRemainingValue * SendPayGasSeparately) - SendIgnoreErrors // Bad
        });
    }

    // Correct usage with variables
    fun correctUsageWithVariables() {
        let modeFlag: Int = SendRemainingValue | SendIgnoreErrors;
        send(SendParameters{
            to: sender(),
            value: 0,
            mode: modeFlag // Ok
        });
    }
    
    fun modeZeroUsage() {
    send(SendParameters{
        to: sender(),
        value: 0,
        mode: 0 // Bad: Should trigger warning about `mode: 0` being redundant
    });
}
}
