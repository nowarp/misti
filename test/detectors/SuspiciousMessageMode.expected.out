[MEDIUM] SuspiciousMessageMode: Flag `SendRemainingValue` is used multiple times in the `mode` expression
test/detectors/SuspiciousMessageMode.tact:16:40:
  15 |             value: 0,
> 16 |             mode: SendRemainingValue | SendRemainingValue // Bad
                                              ^
  17 |         });
Help: Use each flag at most once in the mode expression
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:25:19:
  24 |             value: 0,
> 25 |             mode: SendRemainingValue + SendIgnoreErrors // Bad
                         ^
  26 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Integer literals should not be used in mode expression
test/detectors/SuspiciousMessageMode.tact:34:19:
  33 |             value: 0,
> 34 |             mode: 64 // Bad: Integer literal instead of symbolic constant 
                         ^
  35 |         });
Help: Replace `64` with `SendRemainingValue`
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:52:19:
  51 |             value: 0,
> 52 |             mode: SendRemainingValue + SendRemainingValue + 64 // Bad: Duplicate flags, '+' operator, integer literal 
                         ^
  53 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Flag `SendRemainingValue` is used multiple times in the `mode` expression
test/detectors/SuspiciousMessageMode.tact:52:40:
  51 |             value: 0,
> 52 |             mode: SendRemainingValue + SendRemainingValue + 64 // Bad: Duplicate flags, '+' operator, integer literal 
                                              ^
  53 |         });
Help: Use each flag at most once in the mode expression
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Integer literals should not be used in mode expression
test/detectors/SuspiciousMessageMode.tact:52:61:
  51 |             value: 0,
> 52 |             mode: SendRemainingValue + SendRemainingValue + 64 // Bad: Duplicate flags, '+' operator, integer literal 
                                                                   ^
  53 |         });
Help: Replace `64` with `SendRemainingValue`
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:61:19:
  60 |             value: 0,
> 61 |             mode: -(SendRemainingValue * SendPayGasSeparately) - SendIgnoreErrors // Bad
                         ^
  62 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Mode expression should only contain the '|' operator
test/detectors/SuspiciousMessageMode.tact:61:21:
  60 |             value: 0,
> 61 |             mode: -(SendRemainingValue * SendPayGasSeparately) - SendIgnoreErrors // Bad
                           ^
  62 |         });
Help: Use the '|' operator (bitwise OR) to combine flags
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode

[MEDIUM] SuspiciousMessageMode: Setting `mode` to `0` is redundant as it has no effect.
test/detectors/SuspiciousMessageMode.tact:79:15:
  78 |         value: 0,
> 79 |         mode: 0 // Bad: Should trigger warning about `mode: 0` being redundant
                     ^
  80 |     });
Help: Remove the `mode` field or set it to `SendDefaultMode`.
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousMessageMode